homeassistant/components/subaru/__init__.py

```diff
@@ -0,0 +1,173 @@
+        entry_data["devices"][device.hass_type].append(device)
+
+    for component in SUBARU_COMPONENTS:
+        _LOGGER.debug("Loading %s", component)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in SUBARU_COMPONENTS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    controller = hass.data[DOMAIN][config_entry.entry_id]["controller"]
+    controller.set_update_interval(config_entry.options.get(CONF_HARD_POLL_INTERVAL))
+    controller.set_fetch_interval(config_entry.options.get(CONF_SCAN_INTERVAL))
+
+
+class SubaruDevice(Entity):
+    """Representation of a Subaru Device."""
+
+    def __init__(self, subaru_device, controller, config_entry):
+        """Initialize the Subaru Device."""
+        self.subaru_device = subaru_device
+        self.controller = controller
+        self.config_entry = config_entry
+        self._name = self.subaru_device.name
+        self.subaru_id = slugify(self.subaru_device.uniq_name)
+        self._attributes = {}
+        self._icon = ICONS.get(self.subaru_device.type)
+
+    @property
+    def name(self):
+        """Return name."""
+        return self._name
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self.subaru_id
+
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+        return self._icon
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return self.subaru_device.should_poll
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        attr = self._attributes
+        if self.subaru_device.has_battery():
+            attr[ATTR_BATTERY_LEVEL] = self.subaru_device.battery_level()
+        return attr
+
+    @property
+    def device_info(self):
+        """Return the device_info of the device."""
+        return {
+            "identifiers": {(DOMAIN, self.subaru_device.id())},
+            "name": self.subaru_device.car_name(),
+            "manufacturer": "Subaru",
+        }
+
+    async def async_added_to_hass(self):
+        """Register state update callback."""
+
+    async def async_will_remove_from_hass(self):
+        """Prepare for unload."""
+
+    async def async_update(self):
+        """Update the state of the device."""
+        try:
+            await self.subaru_device.async_update()
+        except SubaruException as ex:
```

Comment:
It looks like you only need to do one update for all entities, consider using this framework with DataUpdateCoordinator https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

Reply:
@MatthewFlamm thanks for the review and comment.  I've been trying to use the DataUpdateCoordinator this week, but I don't think it will work for the current implementation of the subarulink package. It was designed to hide some of the messiness of the Subaru API, especially when dealing with multiple vehicles on one account (it requires multiple API calls to switch vehicle context and then fetch the data update).  The package is caching, so all the extraneous calls for the different entities won't result in any unnecessary API calls being made to Subaru services.  I think the implementation of subarulink.hass needs to eventually go into homeassistant and I'll make that a design goal for a future revision in order to take advantage of the various homeassistant data handling frameworks.

Reply:
Integration has been updated to use DataUpdateCoordinator.  Thanks for the suggestion, this will make the addition of more sensors more efficient.